accepting-cli-commands (#77)
pupilfirst/wd201-scripts: working-with-npm/accepting-cli-commands/README.md

# Text

> In the previous lessons, we learned about how to work with inbuilt features in **Node.js**. While they are very useful in creating applications, the **Node.js** CLI also provides some interesting options to work with user input.

Reviewer: Probably this paragraph can be omitted as it is referring to previous lesson.

Author: This is generic to the lessons earlier as we are using the scripts, streams and HTTP introduced in the previous lessons. I think we can keep this here, as this lesson anyway needs those by default.

Let us learn about how to use the CLI to get user input and use it for our application.

## Accepting input from CLI

Node.js has an inbuilt `readline` module to read input from a readable stream, such as the `process.stdin` stream, which is the one used by the CLI. This stream provides the input during the execution of a **Node.js** program, one line at a time.

Let us write a program to get the user's name and print a greeting for them.

```js
const readline = require("readline");

const lineDetail = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

lineDetail.question(`Please provide your name - `, (name) => {
  console.log(`Hi ${name}!`);
  lineDetail.close();
});
```

The above stream takes the input as a `string` to process the data.

Now, let us try and use the CLI to get user input to write our read content from a file function.

```js
const http = require("http");
const fs = require("fs");
const readline = require("readline");

const lineDetail = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

lineDetail.question(`Please provide the full file path - `, (path) => {

>   const server = http.createServer(function (req, res) {

Reviewer: Creating a server here seems like an overkill. We can just print the file content to the console instead.

Author: Yes. We can. But reusing this as HTTP needs to be introduced at some point, and in the previous lesson they actually learn about it while working on streams.

    const stream = fs.createReadStream(`${path}`);
    stream.pipe(res);
  });
  lineDetail.close();
  server.listen(3000);
});
```

> The above code gets the user input for the file path to the file we want to read, and then streams the result to the HTTP client at `localhost:3000`.

Reviewer: ```suggestion
The above code gets the file path we want to read as the input, and then prints the content of the file on to the console.
```

Author: Not updating for now, based on the above explanation.
